File: js/Simulation.js
    "use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// import { Vector2 } from "three";
// import Common_1 from "./Common.js";
import Controls_1 from "./Controls.js";
import Advection_1 from "./Advection.js";
import ExternalForce_1 from "./ExternalForce.js";
import Viscous_1 from "./Viscious.js";
import Divergence_1 from "./Divergence.js";
import Poisson_1 from "./Poisson.js";
import Pressure_1 from "./Pressure.js";
import { createFbos } from "../utils/createFbos.js";
import Common_1 from "./Common.js"
var Vector2 = THREE.Vector2
var Simulation = /** @class */ (function () {
    function Simulation() {
        this.options = {
            iterations_poisson: 32,
            iterations_viscous: 32,
            mouse_force: 20,
            resolution: 0.5,
            cursor_size: 100,
            viscous: 30,
            isBounce: false,
            dt: 0.014,
            isViscous: false,
            BFECC: true,
        };
        new Controls_1(this.options);
        this.fboSize = new Vector2();
        this.cellScale = new Vector2();
        this.boundarySpace = new Vector2();
        // clacSize
        var width = Math.round(this.options.resolution * Common_1.width);
        var height = Math.round(this.options.resolution * Common_1.height);
        var px_x = 1.0 / width;
        var px_y = 1.0 / height;
        this.cellScale.set(px_x, px_y);
        this.fboSize.set(width, height);
        this.fbos = (0, createFbos)(this.fboSize);
        // createShaderPass()
        this.advection = new Advection_1({
            cellScale: this.cellScale,
            fboSize: this.fboSize,
            dt: this.options.dt,
            src: this.fbos.vel_0,
            dst: this.fbos.vel_1,
        });
        this.externalForce = new ExternalForce_1({
            cellScale: this.cellScale,
            cursor_size: this.options.cursor_size,
            dst: this.fbos.vel_1,
        });
        this.viscous = new Viscous_1({
            cellScale: this.cellScale,
            boundarySpace: this.boundarySpace,
            viscous: this.options.viscous,
            src: this.fbos.vel_1,
            dst: this.fbos.vel_viscous1,
            dst_: this.fbos.vel_viscous0,
            dt: this.options.dt,
        });
        this.divergence = new Divergence_1({
            cellScale: this.cellScale,
            boundarySpace: this.boundarySpace,
            src: this.fbos.vel_viscous0,
            dst: this.fbos.div,
            dt: this.options.dt,
        });
        this.poisson = new Poisson_1({
            cellScale: this.cellScale,
            boundarySpace: this.boundarySpace,
            src: this.fbos.div,
            dst: this.fbos.pressure_1,
            dst_: this.fbos.pressure_0,
        });
        this.pressure = new Pressure_1({
            cellScale: this.cellScale,
            boundarySpace: this.boundarySpace,
            src_p: this.fbos.pressure_0,
            src_v: this.fbos.vel_viscous0,
            dst: this.fbos.vel_0,
            dt: this.options.dt,
        });
    }
    Simulation.prototype.resize = function () {
        var _this = this;
        Object.keys(this.fbos).forEach(function (key) {
            _this.fbos[key].setSize(_this.fboSize.x, _this.fboSize.y);
        });
    };
    Simulation.prototype.update = function () {
        if (this.options.isBounce) {
            this.boundarySpace.set(0, 0);
        }
        else {
            this.boundarySpace.copy(this.cellScale);
        }
        this.advection.updateAdvection(this.options);
        this.externalForce.updateExternalForce({
            cursor_size: this.options.cursor_size,
            mouse_force: this.options.mouse_force,
            cellScale: this.cellScale,
        });
        var vel = this.options.isViscous
            ? this.fbos.vel_1
            : this.viscous.updateViscous({
                viscous: this.options.viscous,
                iterations: this.options.iterations_viscous,
                dt: this.options.dt,
            });
        this.divergence.updateDivergence({ vel: vel });
        var pressure = this.poisson.updatePoisson({
            iterations: this.options.iterations_poisson,
        });
        this.pressure.updatePressure({ vel: vel, pressure: pressure });
    };
    return Simulation;
}());
const _default = Simulation;
export { _default as default };

File: src/Simulation.lua
local window = js.global

local Mouse = require("src.Mouse")
local Common = require("src.Common")
local THREE = window.THREE
local Controls = require("src.Controls")

local Advection = require("src.Advection")
local ExternalForce = require("src.ExternalForce")
local Viscous = require("src.Viscous")
local Divergence = require("src.Divergence")
local Poisson = require("src.Poisson")
local Pressure = require("src.Pressure")
local new = require("utils.new")
local Object = require("utils.convertToJSObject")

local function round(n)
    return math.floor(n+0.5)
end

return function(properties)
    local self = {}
    self.properties = properties

    self.fbos = {
        vel_0 = false,
        vel_1 = false,
        vel_viscous0 = false,
        vel_viscous1 = false,
        div = false,
        pressure_0 = false,
        pressure_1 = false,
    }

    self.options = Object({
        iterations_poisson = 32,
        iterations_viscous = 32,
        mouse_force = 20,
        resolution = 0.5,
        cursor_size = 100,
        viscous = 30,
        isBounce = false,
        dt = 0.014,
        isViscous = false,
        BFECC = true
    })

    Controls(self.options)

    self.fboSize = new(THREE.Vector2)
    self.cellScale = new(THREE.Vector2)
    self.boundarySpace = new(THREE.Vector2)

    self.createAllFBO = function()
        local type = THREE.FloatType

        for k, _ in pairs(self.fbos) do
            self.fbos[k] = new(THREE.WebGLRenderTarget,
                self.fboSize.x,
                self.fboSize.y,
                Object({
                    type = type
                })
            )
        end
    end

    self.createShaderPass = function()
        self.advection = Advection({
            cellScale = self.cellScale,
            fboSize = self.fboSize,
            dt = self.options.dt,
            src = self.fbos.vel_0,
            dst = self.fbos.vel_1
        })

        self.externalForce = ExternalForce({
            cellScale = self.cellScale,
            cursor_size = self.options.cursor_size,
            dst = self.fbos.vel_1,
        })

        self.viscous = Viscous({
            cellScale = self.cellScale,
            boundarySpace = self.boundarySpace,
            viscous = self.options.viscous,
            src = self.fbos.vel_1,
            dst = self.fbos.vel_viscous1,
            dst_ = self.fbos.vel_viscous0,
            dt = self.options.dt,
        })

        self.divergence = Divergence({
            cellScale = self.cellScale,
            boundarySpace = self.boundarySpace,
            src = self.fbos.vel_viscous0,
            dst = self.fbos.div,
            dt = self.options.dt,
        })

        self.poisson = Poisson({
            cellScale = self.cellScale,
            boundarySpace = self.boundarySpace,
            src = self.fbos.div,
            dst = self.fbos.pressure_1,
            dst_ = self.fbos.pressure_0,
        })

        self.pressure = Pressure({
            cellScale = self.cellScale,
            boundarySpace = self.boundarySpace,
            src_p = self.fbos.pressure_0,
            src_v = self.fbos.vel_viscous0,
            dst = self.fbos.vel_0,
            dt = self.options.dt,
        })
    end

    self.calcSize = function()
        local width = round(self.options.resolution * Common.width)
        local height = round(self.options.resolution * Common.height)

        local px_x = 1.0 / width
        local px_y = 1.0 / height

        self.cellScale:set(px_x, px_y)
        self.fboSize:set(width, height)
    end

    self.resize = function()
        self.calcSize()

        for k, v in pairs(self.fbos) do
            self.fbos[k]:setSize(self.fboSize.x, self.fboSize.y)
        end
    end

    self.calcSize()
    self.createAllFBO()
    self.createShaderPass()

    self.update = function()
        if (self.options.isBounce) then
            self.boundarySpace:set(0, 0)
        else
            self.boundarySpace:copy(self.cellScale)
        end

        self.advection.update(self.options)

        self.externalForce.update({
            cursor_size = self.options.cursor_size,
            mouse_force = self.options.mouse_force,
            cellScale = self.cellScale
        })

        local vel = self.fbos.vel_1

        if (self.options.isViscous) then
            vel = self.viscous.update(
                self.options.viscous,
                self.options.iterations_viscous,
                self.options.dt
            )
        end

        self.divergence.update(vel)

        local pressure = self.poisson.update({
            self.options.iterations_poisson
        })

        self.pressure.update(vel, pressure)
    end

    print("Simulation.lua initialized")
    return self
end